speaker/bgp_controller.go

```diff
@@ -272,20 +391,264 @@ type session interface {
+				}
+			case pad.FromAnnotations.Port:
+				port, err := strconv.ParseUint(v, 10, 16)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer port: %v", err)
+				}
+				peerPort = uint16(port)
+			case pad.FromAnnotations.HoldTime:
+				holdTimeRaw = v
+			case pad.FromAnnotations.RouterID:
+				routerID = net.ParseIP(v)
+				if routerID == nil {
+					return nil, fmt.Errorf("invalid router ID %q", v)
+				}
+			}
+		}
+	}
+
+	// Verify required peer config. We shouldn't get errors here because we
+	// validate the configuration. This check is here just for safety.
+	if myASN == 0 {
+		return nil, errors.New("missing local ASN")
+	}
+	if peerASN == 0 {
+		return nil, errors.New("missing peer ASN")
+	}
+	if peerAddr == nil {
+		return nil, errors.New("missing peer address")
+	}
+
+	// Set default BGP port if unspecified by user.
+	if peerPort == 0 {
+		peerPort = 179
+	}
+
+	if holdTime == 0 {
+		// Hold time not specified in autodiscovery defaults - try to parse the
+		// hold time from labels/annotations.
+		ht, err := parseHoldTime(holdTimeRaw)
+		if err != nil {
+			return nil, fmt.Errorf("parsing hold time: %v", err)
+		}
+		holdTime = ht
+	}
+
+	// The peer is configured on a specific node object, so we want to create a
+	// BGP session only on that node.
+	h := ls[v1.LabelHostname]
+	if h == "" {
+		return nil, fmt.Errorf("label %s not found on node", v1.LabelHostname)
+	}
+	ns, err := labels.Parse(fmt.Sprintf("%s=%s", v1.LabelHostname, h))
+	if err != nil {
+		return nil, fmt.Errorf("parsing node selector: %v", err)
+	}
+
+	p := &peer{
+		Cfg: &config.Peer{
+			MyASN:         myASN,
+			ASN:           peerASN,
+			Addr:          peerAddr,
+			Port:          peerPort,
+			HoldTime:      holdTime,
+			RouterID:      routerID,
+			NodeSelectors: []labels.Selector{ns},
+			Password:      password,
+		},
+	}
+
+	return p, nil
+}
+
+// Returns true if the BGP config of a and b is identical.
+//
+// This function compares only BGP configuration - it ignores node selectors,
+// BGP session status and the NodePeer field. It is helpful in cases where we
+// need to check whether two peers are semantically identical with regards to
+// BGP even if they differ in node selectors or if one peer is a node peer and
+// the other isn't.
+//
+// The following parameters are compared: local ASN, peer ASN, peer address,
+// peer port, hold time, router ID. BGP passwords are NOT checked.
+//
+// TODO: When adding BGP password support to node peers, add a check for
+// passwords as well.
+func bgpConfigEqual(a *config.Peer, b *config.Peer) bool {
+	if a.ASN != b.ASN {
+		return false
+	}
+	if !a.Addr.Equal(b.Addr) {
+		return false
+	}
+	if a.MyASN != b.MyASN {
+		return false
+	}
+	if a.Port != b.Port {
+		return false
+	}
+	if a.HoldTime != b.HoldTime {
+		return false
+	}
+	if !a.RouterID.Equal(b.RouterID) {
+		return false
+	}
+
+	return true
+}
+
+func parseHoldTime(ht string) (time.Duration, error) {
```

Comment:
Maybe add a comment saying it is c&p from the function in `internal/config/config.go` and should be kept in sync?

btw, why have you discarded trying to reuse it?

Reply:
Reusing this function requires making it exported in `internal/config`. Also, following https://github.com/metallb/metallb/pull/593#discussion_r439627115 this function is no longer identical to the one in `internal/config` since we now set the hardcoded default at the beginning of `parseNodePeer()` rather than in `parseHoldTime()`.

Reply:
I've looked at it again and decided to export the function in `internal/config`. I've refactored it so that we set the default hold time outside the function in `internal/config`, too. It makes sense to do this anyway since that's the way we deal also with the default BGP port.